### PR TITLE
Optimize Color Parsing

### PR DESCRIPTION
**Pull request by eerhardt**
This method was showing up in a profile of the startup of the .NET Podcasts app.

Optimize it to not allocate intermediate strings on .NET Core, and instead use Span operations.

NOTE: With this change we are not testing the NetStandard build in the unit tests. Should I also add a target for .NET Framework to the unit tests?

### CODE REVIEW

File: src/Microsoft.Maui.Graphics/Color.cs [hunk 1]
```diff
@@ -897,21 +737,295 @@ public static bool TryParse(string value, out Color color)
 			return false;
 		}
 
-		static double ParseColorValue(string elem, int maxValue, bool acceptPercent)
+		static Color GetNamedColor(ReadOnlySpan<char> value)
+		{
+			return s_namedColors.TryGetValue(ComputeIgnoreCaseStringHash(value), out Color color) ? color : null;
+		}
+
+		static Dictionary<uint, Color> s_namedColors = BuildNamedColorDictionary();
```
Comment (jonathanpeppers) [inline src/Microsoft.Maui.Graphics/Color.cs:745]: Is it worth doing:
```suggestion
		static readonly Dictionary<uint, Color> s_namedColors = BuildNamedColorDictionary();
```

File: src/Microsoft.Maui.Graphics/Color.cs [hunk 2]
```diff
@@ -897,21 +737,295 @@ public static bool TryParse(string value, out Color color)
...
+
+		static Dictionary<uint, Color> BuildNamedColorDictionary()
+		{
+			Dictionary<uint, Color> namedColors = new Dictionary<uint, Color>();
```
Comment (jonathanpeppers) [inline src/Microsoft.Maui.Graphics/Color.cs:749]: Can we put the capacity? Is it 143 `AddColor` calls?

**Review by jonathanpeppers — COMMENTED** [review]
Maybe someone familiar with this repo, Is this enough unit tests for this class?

https://github.com/dotnet/Microsoft.Maui.Graphics/blob/main/tests/Microsoft.Maui.Graphics.Tests/ColorUnitTests.cs
https://github.com/dotnet/Microsoft.Maui.Graphics/blob/main/tests/Microsoft.Maui.Graphics.Tests/ColorTypeConverterTests.cs

Just wondering if we'll know if anything breaks.

Should there be a BenchmarkDotNet project in this repo? Maybe that could show the before/after numbers?

**Comment by eerhardt** [comment]
>  Is this enough unit tests for this class?

I beefed up the unit tests for the methods I touched. Please take a look.

> Should there be a BenchmarkDotNet project in this repo? Maybe that could show the before/after numbers?

I added a new one and made a ColorBenchmarker. On my current machine:

```
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19043
Intel Core i7-8700 CPU 3.20GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=6.0.300-preview.22151.9
  [Host]     : .NET Core 6.0.1 (CoreCLR 6.0.121.56705, CoreFX 6.0.121.56705), X64 RyuJIT
  DefaultJob : .NET Core 6.0.1 (CoreCLR 6.0.121.56705, CoreFX 6.0.121.56705), X64 RyuJIT
```

I get the following results:

### Main
|     Method |      Mean |    Error |   StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------- |----------:|---------:|---------:|-------:|------:|------:|----------:|
|      Parse | 101.66 ns | 0.365 ns | 0.341 ns | 0.0267 |     - |     - |     168 B |
| ParseBlack |  59.84 ns | 0.149 ns | 0.124 ns | 0.0050 |     - |     - |      32 B |

### PR
|     Method |     Mean |    Error |   StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------- |---------:|---------:|---------:|-------:|------:|------:|----------:|
|      Parse | 51.88 ns | 0.156 ns | 0.130 ns | 0.0051 |     - |     - |      32 B |
| ParseBlack | 54.26 ns | 0.365 ns | 0.342 ns |      - |     - |     - |         - |

Note that the only allocation left in `Parse` is creating the actual `Color` instance.

**Review by jonathanpeppers — APPROVED** [review]
LGTM 👍

**Comment by jonlipsky** [comment]
@eerhardt This is a great addition!  As you can see from the build, this fails to build on old platforms where Span doesn't exist.  As the main author of the code, other than supporting that case, everything else is there to get this approved and merged.

**Comment by eerhardt** [comment]
@jonlipsky @rmarinho - I have the build working now. I'm not sure what the Bill of Materials step is. But it is failing with:

##[error]The signing feature is not available for your organization yet.

https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=5865234&view=logs&j=02b78222-fc8c-5ab3-e8f3-ab558c416536&t=8880a63a-0c42-57de-c849-ea6ae27cfc7a&l=165

**Comment by jonlipsky** [comment]
@eerhardt Thanks!  I'll be approving shortly!

**Review by jonlipsky — APPROVED** [review]

**Pull request merged**